### PR TITLE
Implement gameplay

### PR DESCRIPTION
**Pull request by ViscousVortex**
Basic gameplay has been added.
 - Player movement.
 - Collisions between player and ceiling/floor.
 - Collisions between player and pillars.

### CODE REVIEW

File: core/src/main/java/org/mini2dx/Tappybird/Background.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package org.mini2dx.Tappybird;
+
+import com.badlogic.gdx.graphics.Texture;
+import org.mini2Dx.core.graphics.Graphics;
+
+import static org.mini2dx.Tappybird.TappyBirdGame.flyingspeed;
+
+public class Background extends Hazards {
+
+    Texture background = new Texture("background.png");
```
Comment (tomcashman) [inline core/src/main/java/org/mini2dx/Tappybird/Background.java]: "background.png" is a magic string and should be moved to a static final variable.

Generally it's a good idea to reuse Texture instances as much as possible. The current implementation will load a new texture and push it to the graphics card every time you make an instance.

File: core/src/main/java/org/mini2dx/Tappybird/InputHandler.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package org.mini2dx.Tappybird;
+
+import com.badlogic.gdx.Gdx;
+import com.badlogic.gdx.Input;
+
+public class InputHandler {
+
+    //TODO Maybe add an abstraction here to the player input.
+    boolean SpacePressed () {
```
Comment (tomcashman) [inline core/src/main/java/org/mini2dx/Tappybird/InputHandler.java]: Method names in Java use the following style:
isSpacePressed()
goToSomewhere()

File: core/src/main/java/org/mini2dx/Tappybird/Player.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        g.drawRect(playerCollisionBox.getX(),playerCollisionBox.getY(),
+                playerCollisionBox.getWidth(),playerCollisionBox.getHeight());
+    }
+
+
+    void update(boolean isJumping){
+
+        if (isJumping){
+            settingPlayerJumping();
+        }
```
Comment (tomcashman) [inline core/src/main/java/org/mini2dx/Tappybird/Player.java]: Styling should follow:

````java
if (condition) {

} else {

}
``

File: core/src/main/java/org/mini2dx/Tappybird/Player.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        //DrawPlayerCollisionBox(g);
+    }
+
+    float getPlayerYPos() {
+        return playerYPos;
+    }
+    float getPlayerYAccel() {
+        return playerYAccel;
+    }
+    float getPlayerTextureHeight() { return playerTextureHeight; }
```
Comment (tomcashman) [inline core/src/main/java/org/mini2dx/Tappybird/Player.java]: Getters should be consistent and use the following style:
```java
float getPlayerTextureHeight() {
   return playerTextureHeight;
}
```

File: core/src/main/java/org/mini2dx/Tappybird/TappyBirdGame.java [hunk 5]
```diff
@@ -1,33 +1,150 @@
...
+        ground2.interpolate(alpha);
     }
     
     @Override
     public void render(Graphics g) {
-		g.drawTexture(texture, 0f, 0f);
+	    fitViewport.apply(g);
+
+	    background1.render(g);
+        background2.render(g);
```
Comment (tomcashman) [inline core/src/main/java/org/mini2dx/Tappybird/TappyBirdGame.java:136]: Try to keep indentation consistent (IntelliJ can format this automatically via Code -> Reformat Code)

File: core/src/main/java/org/mini2dx/Tappybird/TopBottomEdge.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+import com.badlogic.gdx.graphics.Texture;
+import org.mini2Dx.core.graphics.Graphics;
+
+import static org.mini2dx.Tappybird.TappyBirdGame.gameHeight;
+
+public class TopBottomEdge extends Hazards {
+
+    Texture groundTexture = new Texture("groundDirt.png");
+    Texture ceilingTexture = new Texture("ceilingDirt.png");
```
Comment (tomcashman) [inline core/src/main/java/org/mini2dx/Tappybird/TopBottomEdge.java]: Same as above - new texture+ magic string

File: core/src/main/java/org/mini2dx/Tappybird/Player.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import org.mini2Dx.core.graphics.Graphics;
+
+import static org.mini2dx.Tappybird.TappyBirdGame.GRAVITY;
+
+public class Player {
+
+    private static float playerXPos = 100;
+    private static float jumpAccel = -10.0f;
+
+    private Texture playerTexture = new Texture("planeRed1.png");
```
Comment (tomcashman) [inline core/src/main/java/org/mini2dx/Tappybird/Player.java]: Same as above - new texture+ magic string

File: core/src/main/java/org/mini2dx/Tappybird/Player.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    private float playerTextureHeight = playerTexture.getHeight();
+
+    private float playerYPos = 300;
+    private float playerYAccel = 0.0f;
```
Comment (tomcashman) [inline core/src/main/java/org/mini2dx/Tappybird/Player.java:19]: I would change these to playerY and playerYAcceleration

File: core/src/main/java/org/mini2dx/Tappybird/Pillars.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import com.badlogic.gdx.graphics.Color;
+import com.badlogic.gdx.graphics.Texture;
+import org.mini2Dx.core.engine.geom.CollisionBox;
+import org.mini2Dx.core.graphics.Graphics;
+
+
+public class Pillars extends Hazards {
+
+    Texture pillarUp = new Texture("rock.png");
+    Texture pillarDown = new Texture("rockDown.png");
```
Comment (tomcashman) [inline core/src/main/java/org/mini2dx/Tappybird/Pillars.java]: Same as above - new texture+ magic string

**Review by tomcashman — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mini2dx/Tappybird/Background.java (hunk 1 above)
Comment (ViscousVortex) [reply core/src/main/java/org/mini2dx/Tappybird/Background.java]: I implemented a flyweight pattern to stop this from happening.
I generate one instance of a class which contains the shared texture, and I pass a reference to this class where needed.

**Review by tomcashman — APPROVED** [review]
Change the Model classes to Texture classes to be more accurate of what it represents, .i.e. PlayerTexture

**Comment by tomcashman** [comment]
*Change them on next commit

**Pull request merged**